File: 백준/Gold/14719. 빗물/빗물.py
import sys
input = sys.stdin.readline

h,w = map(int,input().split())
arr = [0] + list(map(int,input().split()))
left = [0] * (w + 1)
left[1] = arr[1]
right = [0] * (w + 1)
for i in range(2, w):
    left[i] = max(left[i - 1], arr[i - 1])


right[w] = arr[w]
for i in range(w - 1, 0, -1):
    right[i] = max(right[i + 1], arr[i + 1])
ans = 0
for i in range(2, w):
    tmp = min(left[i],right[i]) - arr[i]
    if tmp <= 0:
        continue
    ans += min(left[i],right[i]) - arr[i]

# print(arr)
# print(left)
# print(right)
print(ans)

File: 백준/Gold/14719. 빗물/빗물.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int h = Integer.parseInt(st.nextToken());
        int w = Integer.parseInt(st.nextToken());
        int[] arr = new int[w + 1];
        int ans = 0;
        int[] prefix = new int[w + 1];
        int[] suffix = new int[w + 1];
        st = new StringTokenizer(br.readLine());
        for (int i = 1; i < w + 1; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }
        for (int i = 2; i < w + 1; i++) {
            prefix[i] = Math.max(arr[i - 1], prefix[i - 1]);
        }

        for (int i = w - 1; i > 0; i--) {
            suffix[i] = Math.max(suffix[i + 1] , arr[i + 1]);
        }

        for (int i = 2; i < w; i++) { //처음 끝 제외
            int tmp = Math.min(prefix[i], suffix[i]) - arr[i];
            if(tmp <= 0) continue;
            ans += tmp;
        }
        System.out.println(ans);

    }
}
